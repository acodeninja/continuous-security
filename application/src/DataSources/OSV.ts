import axios from 'axios';

export class OSV {
  private osvAPIUrl = 'https://api.osv.dev/v1/vulns/';
  private viewVulnerabilityUrl = 'https://osv.dev/vulnerability/';

  private async getData(id: string): Promise<OSVAPIResponse> {
    const response = await axios.get<OSVAPIResponse>(this.osvAPIUrl + id);
    return response.data;
  }

  async getById(id: string): Promise<ReportOutputIssueReference> {
    try {
      const osvData = await this.getData(id);

      let {aliases} = osvData;

      if (osvData.database_specific?.cwe_ids) {
        aliases = aliases.concat(osvData.database_specific.cwe_ids);
      }

      return {
        label: id,
        title: osvData.summary ?? id,
        description: osvData.details,
        directLink: this.viewVulnerabilityUrl + id,
        dataSourceSpecific: {
          osv: {
            aliases,
            severity: osvData.database_specific?.severity?.toLowerCase() as ReportOutputIssueReference['dataSourceSpecific']['osv']['severity'] ?? 'unknown',
          },
        },
      };
    } catch (e) {
      throw new Error(`failed to get id ${id}`);
    }
  }
}
