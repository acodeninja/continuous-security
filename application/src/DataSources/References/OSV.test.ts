import axios from 'axios';
import {OSV} from './OSV';
import {Github} from '../../../tests/fixtures/OSVResponse';

jest.mock('axios', () => ({
  get: jest.fn(),
}));

describe('a GitHub security advisory', () => {
  (axios.get as jest.Mock).mockResolvedValueOnce({data: Github});

  let reference: ReportOutputIssueReference;

  beforeAll(async () => {
    reference = await (new OSV).getById('GHSA-pjpc-87mp-4332');
  });

  test('has the correct label', () => {
    expect(reference.label).toEqual('GHSA-pjpc-87mp-4332');
  });

  test('has the correct title', () => {
    expect(reference.title).toEqual(
      'Cross-site Scripting vulnerability in Mautic\'s tracking pixel functionality',
    );
  });

  test('has the correct description', () => {
    expect(reference.description).toMatchSnapshot();
  });

  test('has the correct url', () => {
    expect(reference.directLink).toEqual('https://osv.dev/vulnerability/GHSA-pjpc-87mp-4332');
  });
});
