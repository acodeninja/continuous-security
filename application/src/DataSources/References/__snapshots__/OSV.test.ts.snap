// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`a GitHub security advisory has the correct description 1`] = `
"### Impact

Mautic allows you to track open rates by using tracking pixels. 
The tracking information is stored together with extra metadata of the tracking request.

The output isn't sufficiently filtered when showing the metadata of the tracking information, which may lead to a vulnerable situation.

### Patches

Please upgrade to 4.3.0

### Workarounds
None.

### References
* Internally tracked under MST-38

### For more information
If you have any questions or comments about this advisory:
* Email us at [security@mautic.org](mailto:security@mautic.org)
"
`;
